### PR TITLE
HPCC-30691 Group functions with related activity in metrics page

### PR DESCRIPTION
**Pull request by GordonSmith**
Currently they can display anywhere in a group on their own

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [ ] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [x] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [x] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

File: esp/src/src-react/util/metricGraph.ts [hunk 1]
```diff
@@ -102,6 +107,7 @@ export class MetricGraph extends Graph2<IScope, IScopeEdge, IScope> {
         super.clear();
         this._index = {};
         this._activityIndex = {};
+        this._subgraphIndex = {};
```
Comment (GordonSmith) [inline esp/src/src-react/util/metricGraph.ts:110]: Not used

File: esp/src/src-react/util/metricGraph.ts [hunk 2]
```diff
@@ -298,15 +312,21 @@ export class MetricGraph extends Graph2<IScope, IScopeEdge, IScope> {
         return "unknown";
     }
 
+    protected _dedupSubgraphs: { [scopeName: string]: boolean } = {};
     subgraphTpl(sg: IScope, options: MetricsOptions): string {
+        if (this._dedupEdges[sg.id] === true) return "";
+        this._dedupEdges[sg.id] = true;
```
Comment (GordonSmith) [inline esp/src/src-react/util/metricGraph.ts:318]: Should be `_dedupSubgraphs`

File: esp/src/src-react/util/metricGraph.ts [hunk 3]
```diff
@@ -298,15 +312,21 @@ export class MetricGraph extends Graph2<IScope, IScopeEdge, IScope> {
...
             childTpls.push(this.subgraphTpl(child, options));
         });
+        if (this.vertexExists(this.id(sg))) {
+            childTpls.push(this.hiddenTpl(this.vertex(this.id(sg)), options));
+        }
         this.subgraphVertices(sg.name).forEach(child => {
             childTpls.push(this.vertexTpl(child, options));
         });
-        this.subgraphEdges(sg.name).forEach(child => {
+        this.subgraphEdges(sg.name).sort((l, r) => l.Label === "LEFT" ? -1 : 0).forEach(child => {
```
Comment (GordonSmith) [inline esp/src/src-react/util/metricGraph.ts:329]: Remove sort

File: esp/src/src-react/util/metricGraph.ts [hunk 4]
```diff
@@ -243,11 +247,18 @@ export class MetricGraph extends Graph2<IScope, IScopeEdge, IScope> {
         return this.outEdges(v.name).filter(e => e.__parentName === v.__parentName);
     }
 
+    protected _dedupVerticies: { [scopeName: string]: boolean } = {};
```
Comment (jeclrsg) [inline esp/src/src-react/util/metricGraph.ts:250]: just a misspelling, should be `_dedupVertices`

File: esp/src/src-react/util/metricGraph.ts [hunk 5]
```diff
@@ -324,6 +342,8 @@ subgraph cluster_${encodeID(sg.id)} {
 
     graphTpl(items: IScope[] = [], options: MetricsOptions) {
         // subgraphs.sort();
```
Comment (jeclrsg) [inline esp/src/src-react/util/metricGraph.ts:344]: commented out line

**Review by jeclrsg — APPROVED** [review]
@GordonSmith marked as approved. Just saw a couple minor things, a commented out line in metricGraph & a typo in a symbol name

**Review by jeclrsg — APPROVED** [review]

File: esp/src/src-react/util/metricGraph.ts [hunk 6]
```diff
@@ -36,8 +40,8 @@ const KindShape = {
     196: "cylinder",        //  Spill Write
 };
 
-function shape(kind: string) {
-    return KindShape[kind] || "rectangle";
+function shape(v: IScope) {
+    return TypeShape[v.type] ?? KindShape[v.kind] ?? "rectangle";
```
Comment (GordonSmith) [inline esp/src/src-react/util/metricGraph.ts:44]: @jeclrsg - should have been `v.Kind`

**Pull request merged**